### ISSUE TITLE
fix(fluoro): backward pass ignores post-render normalisation and inversion, so pose gradients are wrong and the documented 2D/3D registration workflow anti-converges

### ISSUE THREAD

**Issue by chrisvoncsefalvay** [isaac-for-healthcare/i4h-sensor-simulation#68]
## What's wrong

`SlangDiffDRRFunction` stitches together derivatives of two different functions. The forward pass hands PyTorch a normalised, inverted image, computed in numpy outside the autograd graph, while the backward pass applies PyTorch's upstream gradient verbatim to the Slang autodiff of the raw Beer-Lambert intensity. The Jacobian of the post-processing is never applied. With the package defaults (`normalize=True, invert=True`) the returned pose gradient is not merely inaccurate, it is directionally wrong, and the gradient-based 2D/3D registration workflow that the README advertises ("Compute exact gradients for 2D/3D registration", fluoro-simulator/README.md:12) and that the module's own docstrings demonstrate (renderer lines 654 to 665 and 800 to 814) does the opposite of registering. Adam faithfully ascends the loss surface. On a real chest CT the pose error grows from 19.0 mm and 12.4 degrees to 137.6 mm and 118.1 degrees within 200 iterations, while the identical renderer with the post-processing kept inside the graph converges to 0.0005 mm and 0.0009 degrees. In the cleanest sub-case (`invert=True, normalize=False`) the returned gradient is the exact negative of the true one, component for component, to the last bit. 

## Where the boundary breaks

(all references are to commit `945948d`, which is what head is at th time of writing)

The forward pass of the autograd function (`fluorosim/rendering/diffdrr_slang_renderer.py`:570-598) calls `renderer.render()`, which dispatches the Slang forward kernel and then, on the CPU and outside any graph, applies per-frame min-max normalisation (lines 382 to 385) and inversion to the clinical convention (lines 388 to 389). That post-processed array is what `torch.from_numpy` returns to autograd as the graph output (line 598).

The backward pass (lines 600 to 630) takes `grad_output`, which PyTorch computed with respect to that post-processed image, and passes it unmodified to `render_with_gradients` (lines 620 to 625). There it is uploaded as `gradOutput` and consumed by `renderDRR_backward` (`diffdrr_slang.slang`, line 362) as the upstream gradient of `computePixelIntensity`, whose output is the raw intensity `I = i0 * exp(-integral)` (shader line 297). The `[Differentiable]` region ends at the raw intensity; the graph output does not.

Three aggravating observations:

1. Both `normalize` and `invert` default to `True` (`SlangDiffDRRConfig`, renderer lines 133 to 134, mirrored in `XrayPhysics`, `config.py` lines 141 to 142), so every user following the documented examples hits this.
2. `render_with_gradients` has the same inconsistency as a standalone API: it returns the normalised, inverted image (lines 504 to 517) alongside gradients of the raw one (lines 495 to 502), and its own docstring example (lines 426 to 435) computes $\mathtt{grad\_output}=2(\mathtt{synthetic}-\mathtt{target})$ from precisely that returned image.
3. The shader ships a `normalizeImage` kernel (shader lines 391 to 408) that is never dispatched, let alone differentiated. The Python-side numpy normalisation is the one that runs.

## The missing Jacobian

Let $r$ be the raw image, let $\Delta = \max(r) - \min(r)$ and let $\varepsilon$ be the stability constant. Under the defaults the graph output is

$$
f(r) = 1 - \frac{r - \min(r)}{\Delta + \varepsilon}.
$$

Let $J_r = \frac{\partial r}{\partial \theta}$ be the Jacobian of the raw renderer with respect to pose and let $g = \partial L / \partial f$ be the upstream gradient. The backward pass returns $J_r^{\mathsf T}g$. The correct gradient is

$$
\nabla_\theta L = J_r^{\mathsf T}\left(\frac{\partial f}{\partial r}\right)^{\mathsf T}g = -\frac{1}{\Delta + \varepsilon}J_r^{\mathsf T}g + \text{terms through }\min(r)\text{ and }\max(r).
$$

The omitted leading factor is negative: the sign flips. The magnitude error, $1/(\Delta + \varepsilon)$, is largely laundered by Adam's per-parameter preconditioning, which is why the failure presents as clean anti-convergence rather than mere sluggishness. The trailing terms arise because $\min(r)$ and $\max(r)$ are themselves pose-dependent, which also means a constant rescaling of the returned gradient would not fix the defaults case.

With `invert` alone the relationship is exact: $\frac{\partial f}{\partial r} = -I$, so the returned gradient is precisely the negative of the true one. The reproduction below confirms this bitwise.

## How to reproduce

Reproducer is too big for an issue, so it's [here](https://gist.github.com/chrisvoncsefalvay/084c206a1e289aee9c2f2f9bbe1c86d4).

It downloads the 3D Slicer CTChest sample volume (SHA256-pinned), resamples it to 2.5 mm isotropic and maps HU to μ with the package's own `VolumePreprocessor`. The ground truth pose is rotation $\theta^\star=(7,-5,9)^\circ$ and translation $t^\star=(14,-10,8)\,\mathrm{mm}$, a clearly visible misalignment of 12.4 degrees and 19.0 mm; the optimisation starts from zero. The loss is MSE on the rendered images, per the module's own registration example, with Adam at identical hyperparameters for both arms and 200 iterations.

Because the defect lives in the Python-side composition rather than in the kernel, the script isolates it with a line-faithful PyTorch transcription of `computePixelIntensity` (shader line numbers cited throughout the source), replicating both autograd boundary conditions exactly: the current one, labelled current@945948d in the figures (post-processing outside the graph, upstream gradient applied to the raw-intensity VJP, exactly as the repository computes it at that commit) and the corrected one (identical renderer, post-processing inside the graph). This makes the reproduction runnable on any machine. I've included a `--parity` mode that checks the transcription against the live Slang renderer on a CUDA host -- i've put this into a separate flag so it can be de-CId.

Gom8ry follows the package defaults for source-to-detector and source-to-isocentre distances; the march step is 2.5 mm and the detector 128 px, both purely for CPU runtime. I've included two presentations. The volume is oriented so the identity-pose march axis is the anatomical anteroposterior axis, giving a posteroanterior projection rather than a vertex view along the slice stack, and the pixel pitch is 6.0 mm, since the default 512 px at 0.5 mm is a 256 mm plate and hence, at the default magnification of 2, a 128 mm field at the isocentre that crops a thorax to a mediastinal keyhole. Neither choice touches the gradient path, both merely make the evidence legible.

Identical renderer, identical optimiser, identical target and seed. The only difference between the arms is whether the post-processing Jacobian participates in the chain rule. The first arm is labelled current@945948d, the gradient path exactly as the repository computes it at that commit, so the label stays auditable once a fix lands.

| | current@945948d gradient path | corrected gradient path |
|---|---|---|
| final loss (from 2.82e-2) | 3.14e-1 | 1.3e-8 |
| rotation error (from 12.45 deg) | 118.1 deg, having peaked at 120.2 | 0.0009 deg |
| translation error (from 18.97 mm) | 137.6 mm, increasing monotonically | 0.0005 mm |
| recovered pose | (-91.5, 29.4, -46.3) deg, (5.8, -70.0, -115.6) mm | (6.999, -5.000, 9.000) deg, (14.000, -10.000, 8.000) mm |

The image evidence. The current@945948d path has rolled the patient a quarter turn onto their side and differentiated them straight out of the frame in 200 iters:

<img width="1080" height="285" alt="Image" src="https://github.com/user-attachments/assets/c1cb557b-5931-418f-a3b9-58747014c407" />

In the invert-only configuration the current@945948d gradient is the exact negative of the corrected one (`np.array_equal(g_current, -g_corrected)` evals `True`), with cosine similarity to finite differences of -0.991 against +0.991 and sign agreement on 0 of 6 pose components against 6 of 6. Under the full defaults the story is barely different: with the whole thorax and background air in frame the extremal pixels are stable, the min-max Jacobian is dominated by its constant $-1/(\Delta + \varepsilon)$ factor, and the current gradient comes out anti-aligned with the truth at cosine -0.988, against +0.985 for the corrected one, with sign agreement 0 of 6 against 5 of 6, robust across finite-difference step sizes (-0.99 against +0.99 at a 2.5 times coarser step). In narrower fields, where the identity of the extremal pixels switches with pose, the pose-dependent min and max terms add a decorrelating component on top of the sign flip. Either way the returned direction is unusable.

## How to fix it

The best fix for the existing API is to apply the missing vector-Jacobian product at the shared renderer boundary. `render_with_gradients()` already has both quantities required for that calculation: the raw image from its forward dispatch and the upstream gradient with respect to the processed image. It should map that gradient back to raw-intensity space before uploading it to the Slang backward kernel:

```python
# after the fwd dispatch
raw_image = self._output_texture.to_numpy().reshape(height, width)

if raw:
    image = raw_image
else:
    image = _postprocess_image(raw_image, cfg)
    grad_output = _postprocess_vjp(raw_image, grad_output, cfg)

# upload grad_output, now dL/d(raw_image), and dispatch Slang bwd
```

The normalisation VJP must include both the direct $1/(\max(r)-\min(r)+\varepsilon)$ term and the terms through $\min(r)$ and $\max(r)$. Gradients at tied extrema should be shared across the tied pixels, matching `torch.amin` and `torch.amax`. Inversion is the simpler first step: negate the incoming gradient before applying the normalisation VJP.

This placement preserves the current outputs of `render()`, `render_with_gradients()`, the exported `SlangDiffDRRFunction` and `TorchSlangDiffDRR`, while fixing both the standalone NumPy API and the Torch path. A keyword-only `raw=True` option can make the alternative contract explicit: return raw Beer-Lambert intensity and interpret `grad_output` with respect to that raw image.

Moving the display transform into `TorchSlangDiffDRR` is attractive in isolation, but it would make the exported and documented `SlangDiffDRRFunction.apply()` return a different quantity and would still leave the standalone `render_with_gradients()` boundary to repair separately. Applying the VJP at the shared boundary fixes every affected entry point without changing their default observable behaviour.

Two adjacent notes while I'm in the neighbourhood. First, for the differentiable path specifically, per-frame min-max normalisation is a questionable companion to gradient-based registration even once differentiated, both because the switching extremal pixels noted above make the objective only piecewise smooth in pose and for radiometric consistency between frames. a fixed windowing, or a normalisation-invariant loss such as NCC, would serve registration better. I've made this point in my other issue. Second, `backward` currently re-renders the forward pass inside `render_with_gradients`, which doubles the cost of every optimisation step. Caching the forward result in `ctx` would neatly halve it. Happy to file separate issues as this has been bloated enough as is.

Gradient audit, for good measure -- really nicely shows the inversion:

<img width="2468" height="1812" alt="Image" src="https://github.com/user-attachments/assets/55b630a4-d31f-40e9-be3d-e1b96451e986" />
